### PR TITLE
Add spacing adjust mod (reworked)

### PR DESCRIPTION
**Pull request by hlysine**
- Resolves #17128
- Supersedes / closes #17590

This is the branch I originally posted at https://github.com/ppy/osu/pull/17590#issuecomment-1086889449. With the slider rotation branch merged and a few more polishing commits, this branch is now PR ready.

The PR diff looks a bit messy, but it is made up of a number of small patches that don't make sense without the spacing adjust mod implementation for reference, so I've decided to pack them in a single PR.

## Patches to OsuHitObjectGenerationUtils

**b4ae13ef016b7b5be36c289225ad8738bac9bded: Remove the circle radius calculation when clamping objects to playfield**

This is reverting the change made way back in #13688. Having an object inside the playfield means that its center should be inside the playfield borders, not the whole object. This is necessary to prevent the rotation algorithm from triggering in maps with no out-of-bounds objects.

**e494d8d901c45332ab0c5e791b9eccc3c7d50758: Allow objects to stay in place via a config**

Adds `ObjectPositionInfo.StayInPlace` so that spacing adjust can reset HO positions after a break.

**fa9d92928021d82abe6472c710cd60125bca2971: Only rotate objects towards playfield center if they are out of bounds**

This is the key change to make spacing adjust behavior more in line with expectations. Normally with a ranked map + spacing adjust at 1.0x, there won't be any out-of-bounds objects, so the objects are never moved. If an object does go out of bounds, the algorithm will go back 10 objects and start moving them towards the playfield center

------------------

The `OsuModSpacingAdjust` itself is changed to consider anything longer than 1 second as a break and reset HO positions afterward, as per peppy's suggestion in https://github.com/ppy/osu/pull/17590#issuecomment-1087088338.

This PR will also break random mod replays because the generation algorithm has changed.

### CODE REVIEW

**Comment by mk56-spn** [comment]
osu_2022-06-06_01-49-00
There still seems to be issues with sliders not staying within bounds, the mod overall feels like it needs lots of work after playing it extensively, but thats not really relevant to its code quality, so ill leave that for another time

**Comment by hlysine** [comment]
> There still seems to be issues with sliders not staying within bounds

That seems to be an issue introduced in slider rotation PR (#17591), which means it affects the random mod as well. I'll fix it in a separate pull when I have time to get to it.

> the mod overall feels like it needs lots of work after playing it extensively, but thats not really relevant to its code quality, so ill leave that for another time

Where exactly do you think the mod needs work? I would love suggestions on ways to improve this implementation. I think it's fine to discuss here as long as it is related to the code presented here.

**Comment by mk56-spn** [comment]
alright ill give my suggestions then 

**Preservation of general map aesthetic**
I've found that very high scaling SA value deform the map ,obviously , it only make sense that they would , but i think it goes too far when moving things like 1-2 jumps and spaced streams, if jumps are vertical in the original map they should IMO remain so with spacing increase, lest all maps end up being diagonal jump ish, which i have found comes close to being a thing on many

https://streamable.com/vwnr7u

**Naming of values**

i don't think calling the spacing adjustment .5 and 2.0 is very user friendly in thgis case since unlike something like DT the correlation isn't as accurate in the map, i think more vague terms such as low spacing and high spacing would be more fitting

**Issues with stacked notes**
It seems that when notes are stacked over a certain amount the mod seems to refuse to separate them no matter how high the spacing gets, whilst being overkill the second the original notes are slightly spaced apart
osu_2022-06-06_16-09-14
kimi no kioku with 2x spacing still has overlapping doubles
osu_2022-06-06_16-10-49=
whilst  plasma gun becomes this, and yes i think there should be a noticeable difference since they start off with different spacing but i think you should lower the effect on the higher end and increase it on the lower end

**Poor readability**
osu_2022-06-06_16-13-26
when SA is at its minimum value it tends to create very hard to read overlaps  with Sliders and circles, i think a bandaid solution for this is to make sliders that were not overlapping a circle at normal spacing not overlap at any value, but only with sliders so streams  can still reduce their spacing adequately

I have other concerns such as the usefulness of the mod for specific things if it does not get expanded ( an option to only space object above or below a certain BPM would be very nice for stream and jump heavy players) but those aren't pertinent right now, overall i think this is a great addition though and look forward to seeing it added

**Comment by hlysine** [comment]
Thanks for the suggestions. I've done a lot of playtesting since then and I agree with most of your points. But instead of applying a bunch of band-aids, I've decided to use a non-linear scaling so that the farther apart two circles are, the less they will be pulled apart. This should hopefully reduce the number of diagonal jumps. On the other hand, short jumps get less of a distance reduction than longer jumps when object spacing < 1, hopefully increasing readability. You can see a graph of the formula here: https://www.desmos.com/calculator/n9eotbynwv

> Naming of values
> 
> i don't think calling the spacing adjustment .5 and 2.0 is very user friendly in thgis case since unlike something like DT the correlation isn't as accurate in the map, i think more vague terms such as low spacing and high spacing would be more fitting

I understand that 0.5x-2.0x might be misleading because it is not accurate at all, but just "low/high spacing" would be too vague and would not represent a continuous range well. I don't have any good ideas for this, so I've kept it as is.

> Issues with stacked notes
> It seems that when notes are stacked over a certain amount the mod seems to refuse to separate them no matter how high the spacing gets, whilst being overkill the second the original notes are slightly spaced apart

Notes can't be separated if they were originally stacked pixel-perfectly because we have no information on which direction the notes should separate to. But as long as they are slightly apart, the spacing adjustment can be applied normally. I don't think this is a problem though. Stacked notes should remain stacked no matter what.

**Comment by mk56-spn** [comment]
thanks for incorporating the changes. played it. feels better, mergeable imo

**Comment by respektive** [comment]
Played around with this for a bit and it feels pretty nice now imo. Hoping to see this merged soon :+1:

**Comment by peppy** [comment]
I dunno about this. It doesn't look like the above issues are improved much. Testing with disco prince, it changes the angles of hitobjects when they shouldn't need to be changed, and the 0.5-2.0 range looks nothing like what i'd expect from those numbers.

**Comment by hlysine** [comment]
> it changes the angles of hitobjects when they shouldn't need to be changed

Angles need to be changed aggressively to avoid objects clumping together in a corner of the playfield, which is very common when you only scale up the distance but leave the angles unchanged.

> the 0.5-2.0 range looks nothing like what i'd expect from those numbers.

That's true. There are so many factors affecting the actual distance that it rarely follows the selected scale. Ideally this setting should offer a continuous range without implying a direct scaling, but I can't think of a better way to show it.

--------------------------------

Ultimately, the algorithm is more concerned about producing a playable map than a map that looks like the original one scaled directly. Instead of changing the algorithm to preserve original jump angles (which I think is impossible), do you think it's acceptable to just rename this mod to something like "Remapped" to remove that expectation?

**Comment by mk56-spn** [comment]
Theres ways more of the maps feel could be kept without anything drastic that would affect playability too much , like allowing sliders to move but not rotate themelves off of their original angles , especially in grid maps, this would help immensely in preserving feel, and forcing circles on said maps to remain at the possible angles for that sort of map

**Comment by hlysine** [comment]
Slider rotation was added to increase playability because there were often awkward angle changes between a slider head and the preceding object when the slider didn't rotate, so I'm not sure reverting that change would help.

----------------

I've done another iteration of tweaking and playtesting. This time, the backtracking distance of the algorithm is drastically reduced. This has several benefits:

- most parts of maps are directly scaled, which is especially good for grid maps
- since there is more direct scaling, the 0.5x-2.0x setting makes more sense now

However, less backtracking means objects are more likely to clump together near the playfield edges and some angle changes are more abrupt. 

Please have a go at this and tell me how this iteration feels, especially on maps with more structured object placement.

----------------

By the way, having the spacing default set at 1.0x feels pretty bad to me, since 1.0x is the only value that doesn't make much sense. Should I split this mod into Spacing Increase + Spacing Decrease? It would work like Half Time + Double Time I think.

**Comment by mk56-spn** [comment]
Tested, as you said it improves some issues at the cost of slightly worsening others

but anyways ill say my piece, i think this should be merged so it gains visibility and can be iterated upon quicker with more eyes on it, its in a decent to good state on many maps ( especially some stream maps and large jump ones, and i have ideas for fixing disco princ- esque maps, tech maps etc... but it'll take time and probably a rewrite later down the line) IMO we would do best to treat this and random like PP in a way , a work in progress, although i don't know how open ppy and the other devs are  to the idea of an ever morphing mod.

tldr: imo fit to be integrated and be built upon later.

**Comment by mk56-spn** [comment]
as for the splitting it, no opinion on it, although splitting diff increase and decrease  (baring DA) seems to be the precedent so shrug
